adr/ADR-15.md

```diff
@@ -0,0 +1,138 @@
+
+If the result indicates that the consumer already exists, then it means that there is an active durable consumer, with some NATS subscription attached and so we should fail this, *unless* a Queue name was provided. The client/server currently does not use the concept of a queue group in the JS consumer, but if the server is enhanced, as part of getting the consumer information, the library will be able to determine if the consumer that already exists is part of the same group, in which case it is fine, or not, in which case an error should be returned.
+
+When the consumer already exists, the NATS queue subscription that was created prior to the "AddConsumer" call needs to be destroyed and replaced with the new NATS queue subscription on the consumer info's `DeliverSubject`.
+
+### Runtime
+
+#### Heartbeats
+
+When heartbeats have been configured, the server will send heartbeats to the client. A timer should be set to report through the asynchronous error callback if heartbeats are detected as missed.
+
+In hearbeats mode, the library should keep track of all messages which reply subject start with `$JS.ACK` since they contain information about the stream and consumer sequence, etc..
+
+When the library receives a message that is found to be a control message (no payload and containst the "Status" header), if the header is "100 Idle Heartbeat", then the library should check if there is any gap detected between what was the last tracked message meta information and the content of the heartbeat. If a gap is detected, an asynchronous error should be reported indicating what the gap is.
+
+Ordered consumers (should have its own ADR) will handle all that for the application, recreating the subscription as needed at the right stream sequence.
+
+#### Flow control
+
+When flow control is enabled, the library receives a control message that indicates that this is a flow control and should respond to the provided subject when the current pending message has been processed.
+
+It is possible, however, that either the flow control or its response is missed and that the consumer is considered stalled from a server perspective.
+
+The server includes in the heartbeats a header to indicate if the consumer is stalled and a subject to send an empty message to. When that happens, the library need to respond to this message, regardless of its internal state of flow control.
+
+Without this, it can be that a consumer remains stalled, so having both heartbeats and flow control configured seem like necessary, so we may make that a misconfiguration if flow control is configured without heartbeat.
+
+Again, Ordered consumers should be favored.
+
+### Unsubscribe and Drain
+
+When a subscription is unsubscribed or drained, if the library created the JS consumer, (or it already existed and this is a queue member added to the group), at the end of those calls the JS consumer should be deleted.
```

Comment:
If we joined an already-existing consumer as a queue member, shouldnt we leave it alone and not delete?